typescript/infra/src/core/govern.ts

```diff
@@ -0,0 +1,109 @@
+import { ethers } from 'ethers';
+
+import {
+  AbacusCoreChecker,
+  ChainMap,
+  ChainName,
+  CoreViolationType,
+  OwnerViolation,
+  ValidatorViolation,
+  ValidatorViolationType,
+  ViolationType,
+  objMap,
+} from '@abacus-network/sdk';
+import { types } from '@abacus-network/utils';
+
+interface Call {
+  to: types.Address;
+  data: ethers.utils.BytesLike;
+}
+
+export class AbacusCoreGovernor<Chain extends ChainName> {
+  readonly checker: AbacusCoreChecker<Chain>;
+  calls: ChainMap<Chain, Call[]>;
+
+  constructor(checker: AbacusCoreChecker<Chain>) {
+    this.checker = checker;
+    this.calls = objMap(this.checker.app.contractsMap, () => []);
+  }
+
+  pushCall(chain: Chain, call: Call) {
+    this.calls[chain].push(call);
+  }
+
+  async govern() {
+    for (const violation of this.checker.violations) {
+      switch (violation.type) {
+        case CoreViolationType.Validator: {
+          await this.handleValidatorViolation(violation as ValidatorViolation);
+          break;
+        }
+        case ViolationType.Owner: {
+          await this.handleOwnerViolation(violation as OwnerViolation);
```

Comment:
surprised you need to cast here
there's definitely a "type discrimination" approach here that the compiler should understand but nbd

Reply:
https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions